### PR TITLE
Java: Factor NumberFormatException out into a library file

### PR DESCRIPTION
**Pull request by luchua-bc**
Convert the query <code>src/Violations of\ Best Practice/Exception Handling/NumberFormatException.ql</code> to a library so that it can shared by all the queries dealing with NFEs.

### CODE REVIEW

File: java/ql/src/semmle/code/java/NumberFormatException.qll [hunk 1]
```diff
@@ -0,0 +1,83 @@
+/**
```
Comment (intrigus-lgtm) [inline java/ql/src/semmle/code/java/NumberFormatException.qll:1]: This is not needed as this is not a query but a library.

File: java/ql/src/semmle/code/java/NumberFormatException.qll [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+    this.isStringConstructor("Double")
+  }
+}
+
+/** The class `java.lang.NumberFormatException` */
+class NumberFormatException extends RefType {
+  NumberFormatException() { this.hasQualifiedName("java.lang", "NumberFormatException") }
+}
+
+/** Holds if NFE is caught */
```
Comment (Marcono1234) [inline java/ql/src/semmle/code/java/NumberFormatException.qll:71]: Could `NumberFormatException` be written out here? I think that would make it more clearer than "NFE".

File: java/ql/src/semmle/code/java/NumberFormatException.qll [hunk 3]
```diff
@@ -0,0 +1,83 @@
...
+predicate catchesNFE(TryStmt t) {
+  exists(CatchClause cc, LocalVariableDeclExpr v |
+    t.getACatchClause() = cc and
+    cc.getVariable() = v and
+    v.getType().(RefType).getASubtype*() instanceof NumberFormatException
+  )
+}
+
+/** Holds if NFE is thrown */
```
Comment (Marcono1234) [inline java/ql/src/semmle/code/java/NumberFormatException.qll:80]: Could `NumberFormatException` be written out here? I think that would make it more clearer than "NFE".

File: java/ql/src/semmle/code/java/NumberFormatException.qll (hunk 1 above)
Comment (luchua-bc) [reply java/ql/src/semmle/code/java/NumberFormatException.qll]: Thanks @intrigus-lgtm for reviewing this PR. I've made the change.

File: java/ql/src/semmle/code/java/NumberFormatException.qll (hunk 2 above)
Comment (luchua-bc) [reply java/ql/src/semmle/code/java/NumberFormatException.qll]: Updated

File: java/ql/src/semmle/code/java/NumberFormatException.qll (hunk 3 above)
Comment (luchua-bc) [reply java/ql/src/semmle/code/java/NumberFormatException.qll]: Updated

File: java/ql/src/semmle/code/java/NumberFormatException.qll [hunk 4]
```diff
@@ -0,0 +1,71 @@
+import java
```
Comment (aschackmull) [inline java/ql/src/semmle/code/java/NumberFormatException.qll:1]:
```suggestion
/** Provides classes and predicates for reasoning about `java.lang.NumberFormatException`. */
import java
```

File: java/ql/src/semmle/code/java/NumberFormatException.qll [hunk 5]
```diff
@@ -0,0 +1,71 @@
...
+
+/** Calls a string to number conversion */
```
Comment (aschackmull) [inline java/ql/src/semmle/code/java/NumberFormatException.qll:3]:
```suggestion
/** A call to a string to number conversion. */
```

File: java/ql/src/semmle/code/java/NumberFormatException.qll [hunk 6]
```diff
@@ -0,0 +1,71 @@
...
+    this.isValueOfMethod("Byte") or
+    this.isValueOfMethod("Short") or
+    this.isValueOfMethod("Integer") or
+    this.isValueOfMethod("Long") or
+    this.isValueOfMethod("Float") or
+    this.isValueOfMethod("Double")
+  }
+}
+
+/** Constructs a number from its string representation */
```
Comment (aschackmull) [inline java/ql/src/semmle/code/java/NumberFormatException.qll:36]:
```suggestion
/** A `ClassInstanceExpr` that constructs a number from its string representation. */
```

File: java/ql/src/semmle/code/java/NumberFormatException.qll [hunk 7]
```diff
@@ -0,0 +1,71 @@
...
+    this.isStringConstructor("Byte") or
+    this.isStringConstructor("Short") or
+    this.isStringConstructor("Integer") or
+    this.isStringConstructor("Long") or
+    this.isStringConstructor("Float") or
+    this.isStringConstructor("Double")
+  }
+}
+
+/** The class `java.lang.NumberFormatException` */
```
Comment (aschackmull) [inline java/ql/src/semmle/code/java/NumberFormatException.qll:54]:
```suggestion
/** The class `java.lang.NumberFormatException`. */
```

File: java/ql/src/semmle/code/java/NumberFormatException.qll [hunk 8]
```diff
@@ -0,0 +1,71 @@
...
+class NumberFormatException extends RefType {
+  NumberFormatException() { this.hasQualifiedName("java.lang", "NumberFormatException") }
+}
+
+/** Holds if `java.lang.NumberFormatException` is caught */
```
Comment (aschackmull) [inline java/ql/src/semmle/code/java/NumberFormatException.qll:59]:
```suggestion
/** Holds if `java.lang.NumberFormatException` is caught. */
```

File: java/ql/src/semmle/code/java/NumberFormatException.qll [hunk 9]
```diff
@@ -0,0 +1,71 @@
...
+predicate catchesNFE(TryStmt t) {
+  exists(CatchClause cc, LocalVariableDeclExpr v |
+    t.getACatchClause() = cc and
+    cc.getVariable() = v and
+    v.getType().(RefType).getASubtype*() instanceof NumberFormatException
+  )
+}
+
+/** Holds if `java.lang.NumberFormatException` is thrown */
```
Comment (aschackmull) [inline java/ql/src/semmle/code/java/NumberFormatException.qll:68]:
```suggestion
/** Holds if `java.lang.NumberFormatException` can be thrown. */
```

**Review by aschackmull — CHANGES_REQUESTED** [review]
Some qldoc changes, otherwise LGTM.

File: java/ql/src/semmle/code/java/NumberFormatException.qll (hunk 4 above)
Comment (luchua-bc) [reply java/ql/src/semmle/code/java/NumberFormatException.qll:3]: Thanks @aschackmull. I've updated the qldoc.

File: java/ql/src/semmle/code/java/NumberFormatException.qll (hunk 5 above)
Comment (luchua-bc) [reply java/ql/src/semmle/code/java/NumberFormatException.qll]: Updated

File: java/ql/src/semmle/code/java/NumberFormatException.qll (hunk 6 above)
Comment (luchua-bc) [reply java/ql/src/semmle/code/java/NumberFormatException.qll]: Updated

File: java/ql/src/semmle/code/java/NumberFormatException.qll (hunk 7 above)
Comment (luchua-bc) [reply java/ql/src/semmle/code/java/NumberFormatException.qll]: Updated

File: java/ql/src/semmle/code/java/NumberFormatException.qll (hunk 8 above)
Comment (luchua-bc) [reply java/ql/src/semmle/code/java/NumberFormatException.qll]: Updated

File: java/ql/src/semmle/code/java/NumberFormatException.qll (hunk 9 above)
Comment (luchua-bc) [reply java/ql/src/semmle/code/java/NumberFormatException.qll]: Updated

**Review by aschackmull — APPROVED** [review]

**Comment by luchua-bc** [comment]
Thanks @aschackmull for all your help with this PR.

**Pull request merged**